types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,346 @@
+		.button = button,
+		.state = state ? WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.button, &event);
+}
+
+static void virtual_pointer_axis(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t axis,
+		wl_fixed_t value) {
+	if (axis > WL_POINTER_AXIS_HORIZONTAL_SCROLL) {
+		wl_resource_post_error(resource,
+				ZWLR_VIRTUAL_POINTER_V1_ERROR_INVALID_AXIS,
+				"Invalid enumeration value %d", axis);
+		return;
+	}
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis = axis;
+	pointer->axis_event[pointer->axis].device = wlr_dev;
+	pointer->axis_event[pointer->axis].time_msec = time;
+	pointer->axis_event[pointer->axis].orientation = axis;
+	pointer->axis_event[pointer->axis].delta = wl_fixed_to_double(value);
+}
+
+static void virtual_pointer_frame(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+
+	if (pointer->axis_event[pointer->axis].device) {
+		/* Deliver pending axis event */
+		wlr_signal_emit_safe(&wlr_dev->pointer->events.axis,
+				&pointer->axis_event[pointer->axis]);
+		memset(&pointer->axis_event[pointer->axis],
+				0, sizeof(pointer->axis_event[pointer->axis]));
+	}
+
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.frame, wlr_dev->pointer);
+}
+
+static void virtual_pointer_axis_source(struct wl_client *client,
+		struct wl_resource *resource, uint32_t source) {
+	if (source > WL_POINTER_AXIS_SOURCE_WHEEL_TILT) {
+		wl_resource_post_error(resource,
+				ZWLR_VIRTUAL_POINTER_V1_ERROR_INVALID_AXIS_SOURCE,
+				"Invalid enumeration value %d", source);
+		return;
+	}
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis_event[pointer->axis].device = wlr_dev;
+	pointer->axis_event[pointer->axis].source = source;
+}
+
+static void virtual_pointer_axis_stop(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t axis) {
+	if (axis > WL_POINTER_AXIS_HORIZONTAL_SCROLL) {
+		wl_resource_post_error(resource,
+				ZWLR_VIRTUAL_POINTER_V1_ERROR_INVALID_AXIS,
+				"Invalid enumeration value %d", axis);
+		return;
+	}
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis = axis;
+	pointer->axis_event[pointer->axis].device = wlr_dev;
+	pointer->axis_event[pointer->axis].time_msec = time;
+	pointer->axis_event[pointer->axis].orientation = axis;
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis,
```

Comment:
I think we should set `delta`/`delta_discrete` to zero and let the `frame` request handler send the event. This will prevent two `axis` events for the same axis from being sent during the same frame.